### ISSUE TITLE
Can/should Bootstrap.serve take a Finagle Filter?

### ISSUE THREAD

**Issue by vkostyukov** [finagle/finch#1051]
This is a very popular use-case: add a filter to certain endpoints wile leaving others. Here is the recent request [about auth](https://gitter.im/finagle/finch?at=5c0bda3126de6f0822a1924c).

Not sure about how API would look like and would love to hear others thoughts on this.

The sketch may look as:

```scala
Bootstrap
  .serveFiltered[Appication.Json](authFilter, helloEndpoint)
```

This means, in addition to keeping track of endpoints in an `HList`, we also need to keep track of filters.

**Comment by sergeykolbasov** [comment]
I'm curious about semantics. So the given filter should be applied to the whole `Service` output, or only to the endpoints, it was mentioned with?

**Comment by prayagupd** [comment]
Hi, my use case was all of the API endpoints are secured(`api/*`), except two which are`/heartbeat` (gives API health) and `/health` (overall health including connectivity to external dependent APIs), which will be pinged by kubernetes. (For ops purpose basically)

So, the secured endpoints are intercepted with the filter and make sure the clients are authorized to access. (Using `api-secret` http header)

My current solution is using [`finagle.http.HttpMuxer`](https://twitter.github.io/finagle/docs/com/twitter/finagle/http/HttpMuxer.html),

```scala
  val heartbeat: Endpoint[String] = get("heartbeat") {
    NoContent[String]
  }

  val healthcheck: Endpoint[Map[String, Int]] = get("healthcheck") {
    Ok(
      Map(
        "service1" -> 200,
        "service2" -> 200
      )
    )
  }

 val secureEndpoints = aEndpoint :+: bEndpoint :+: cEndpoint

  val allEndpoints = HttpMuxer
    .withHandler("api/", EndpointOps.authFilter.andThen(secureEndpoints.toService))
    .withHandler("heartbeat", heartbeat.toService)
    .withHandler("health", healthcheck.toService)

Await.ready(Http.server.serve(":8080", allEndpoints))
```

I was thinking if Service could be thought as Semigroup so that they can be combined but I'm not sure. 
The `.serve` that accepts filter can be good one as well, so that above solution can work as

```scala
Bootstrap
  .serve[Appication.Json](authFilter, (aEndpoint :+: bEndpoint :+: cEndpoint))
  .serve[Application.Json](healthEndpoint)
  .serve[Text.Plain](heatbeatEndpoint)
```

**Comment by sergeykolbasov** [comment]
In my opinion, it's not that simple and I think we need to set up the common ground first.

Here is the logic on transforming endpoints to Finagle services:
https://github.com/finagle/finch/blob/master/core/src/main/scala/io/finch/ToService.scala#L102

1: Should we call any endpoint before applying a filter, or 2: should we just pass the resulting `Service` to the given filter?

The problem is the different behavior for those cases:

- In the first case we call `Endpoint.apply` and depending on customer endpoints, things might got called. It's the same as using custom Endpoint for performing auth. We don't evaluate the output yet, but the endpoints are applicative, so we execute the matching logic anyway.

- In the second case, I see no possibility to separate filter application per endpoint groups, because for each `.serve` we resolve the `Service` instance and call it from the previously resolved `Service` until we reach the end of the list and return 4xx error.

In the given example:
```scala
Bootstrap
  .serve[Appication.Json](authFilter, (aEndpoint :+: bEndpoint :+: cEndpoint)) //authFilter will be applied to the rest of endpoints: health & heartbeat
  .serve[Application.Json](healthEndpoint)
  .serve[Text.Plain](heatbeatEndpoint)
```

There is no sane & overhead-free way to build list of separate Services w/wo Filters and call them one by one, as we don't know how to distinguish between response from Finch and customer endpoint.

I've been working on this for quite a while already, and the root cause I think is that Endpoint is Applicative, but not a Monad. It was discussed already in #649. The way I see how to solve the initial problem is to have an option to stop the evaluation of endpoints matching once failure occurs, so to have a monadic composition with fail-fast logic. But it's impossible with the current API.

**Comment by sergeykolbasov** [comment]
Closed in favor of #1080
